docs/specs/buffer.md

```diff
@@ -18,6 +18,16 @@ Vector buffers MUST be instrumented for optimal observability and monitoring. Th
 
 ### Events
 
+#### `BufferCreated`
+
+*All buffers* MUST emit a `BufferCreated` event immediately upon creation.
+
+* Properties
+  * `max_size` - the maximum number of events or byte size of the buffer
+  * `id` - the ID of the component associated with the buffer
+* Metric
+  * MUST emit the `buffer_max_event_size` gauge (in-memory buffers) or `buffer_max_byte_size` gauge (disk buffers) with the defined `max_size` value and `id` as tag.
```

Comment:
For disk buffers, should this also emit the initial counters for the number of events that were stored on disk?

We will likely also need to publish this max on a regular interval to make sure the timestamp is always fresh (systems like Prometheus will reject metrics with stale timestamps).

Reply:
Ah, great point. Yes, I think emitting bytes of events stored on disk makes sense -- not immediately sure how to achieve number of events. However, as mentioned to me by @blt, we may unify buffers under the byte unit in the future.

For emitting at regular interval, is there a convention/example for this? I'm not sure where it makes sense to continue publishing.

Reply:
We do have a place where we scan the disk buffer on start-up so we should be able to publish the number of events there too, perhaps:

https://github.com/vectordotdev/vector/blob/master/lib/vector-core/buffers/src/disk/leveldb_buffer/mod.rs#L52-L59

For emitting on a regular interval, @blt or @tobz might have thoughts, but I would think you could just use a tokio timer.

Reply:
Basically a separate task with a timer or something, yeah.  You could theoretically clone the `Arc<AtomicUsize>` we use for "current size" in disk buffers, and do roughly the same for the in-memory buffers, and then hand that off to your reporting task.  Trickiest thing is preserving tags, but that could likely be handled pretty easily e.g. by wrapping the task spawn in a span when building it, since you should be in the same area where you have access to the sink component name/type/ID/etc.

Reply:
Thanks! Will look into using tokio for recurring `max_size` metric. 

For the initial `received_events/bytes_total` counters, I realize emitting around the point of building the underlying leveldb instance ends up lacking the span context that eventually tags the counters. Unless we end up passing more component metadata into buffer construction, I believe it's only aware of `component_id` as part of the buffer path. Perhaps figuring out the gauge issue from before will help me here; let me know if I have any misunderstandings.

Reply:
👍 at the very least, `component_id` should be sufficient for users. If we can inject the other span context, even better though.